#11164 got master broken

<!-- Thank you for reporting a bug to Coq! -->

#### Description of the problem

On [our CI](https://gitlab.mpi-sws.org/iris/iris/-/jobs/52379), we are seeing build failures for the coq.dev opam package:
```
165 #=== ERROR while compiling coq.dev ============================================#
166 # context              2.0.5 | linux/x86_64 | ocaml-base-compiler.4.07.1 | pinned(git+https://github.com/coq/coq.git#master#a9a06ffb)
167 # path                 /builds/iris/iris/opamroot/ocaml-base-compiler.4.07.1/.opam-switch/build/coq.dev
168 # command              /usr/bin/make -j20
169 # exit-code            2
170 # env-file             /builds/iris/iris/opamroot/log/coq-303-d6d332.env
171 # output-file          /builds/iris/iris/opamroot/log/coq-303-d6d332.out
172 ### output ###
173 # [...]
174 # OCAMLOPT  vernac/recLemmas.ml
175 # OCAMLOPT  plugins/firstorder/unify.ml
176 # OCAMLOPT  user-contrib/Ltac2/tac2extffi.ml
177 # File "pretyping/recordops.ml", line 234, characters 28-31:
178 # Error: Unbound value con
179 # Hint: Did you mean cos?
180 # Makefile.build:747: recipe for target 'pretyping/recordops.cmx' failed
```

I can reproduce this failure locally with ocaml-system.4.08.1.

#### Coq Version

<!-- You can get this information by running `coqtop -v`. If relevant, please
also include your operating system. -->
git master

--- a/pretyping/recordops.ml
+++ b/pretyping/recordops.ml
@@ -138,7 +138,7 @@ let find_primitive_projection c =
 *)
 
 type obj_typ = {
-  o_ORIGIN : Constant.t;
+  o_ORIGIN : GlobRef.t;
   o_DEF : constr;
   o_CTX : Univ.AUContext.t;
   o_INJ : int option;      (* position of trivial argument if any *)
@@ -212,7 +212,8 @@ let compute_canonical_projections env ~warn (gref,ind) =
         mkConstU (con, u), Environ.constant_value_in env (con,u)
     | GlobRef.VarRef id ->
         mkVar id, Option.get (Environ.named_body id env)
-    | GlobRef.ConstructRef _ | GlobRef.IndRef _ -> assert false in
+    | GlobRef.ConstructRef _ | GlobRef.IndRef _ -> assert false
+  in
   let sign,t = Reductionops.splay_lam env (Evd.from_env env) (EConstr.of_constr c) in
   let sign = List.map (on_snd EConstr.Unsafe.to_constr) sign in
   let t = EConstr.Unsafe.to_constr t in
@@ -231,7 +232,7 @@ let compute_canonical_projections env ~warn (gref,ind) =
             match cs_pattern_of_constr nenv t with
             | patt, o_INJ, o_TCOMPS ->
               ((GlobRef.ConstRef proji_sp, (patt, t)),
-               { o_ORIGIN = con ; o_DEF ; o_CTX ; o_INJ ; o_TABS ; o_TPARAMS ; o_NPARAMS ; o_TCOMPS })
+               { o_ORIGIN = gref ; o_DEF ; o_CTX ; o_INJ ; o_TABS ; o_TPARAMS ; o_NPARAMS ; o_TCOMPS })
               :: acc
             | exception Not_found ->
               if warn then warn_projection_no_head_constant (sign, env, t, gref, proji_sp);

--- a/pretyping/recordops.mli
+++ b/pretyping/recordops.mli
@@ -73,7 +73,7 @@ type cs_pattern =
   | Default_cs
 
 type obj_typ = {
-  o_ORIGIN : Constant.t;
+  o_ORIGIN : GlobRef.t;
   o_DEF : constr;
   o_CTX : Univ.AUContext.t;
   o_INJ : int option;      (** position of trivial argument *)

--- a/vernac/prettyp.ml
+++ b/vernac/prettyp.ml
@@ -1003,9 +1003,7 @@ let print_canonical_projections env sigma grefs =
       | Const_cs y -> GlobRef.equal y gr
       | _ -> false
     end ||
-    match gr with
-    | GlobRef.ConstRef con -> Names.Constant.equal c.o_ORIGIN con
-    | _ -> false
+    GlobRef.equal c.o_ORIGIN gr
   in
   let projs =
     List.filter (fun p -> List.for_all (match_proj_gref p) grefs)

